### DISCUSSION TITLE
Basic question related to coupling in kernels

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#21796] [Q&A Getting Started] [answered]
Hello,
I have a very basic question related to ` params.addRequiredCoupledVar` in kernels.
I am currently passing more variables than what actually being used by the kernel using this param. I briefly looked into the code for it and it seems like it's accumulating dependency list, so I am imagining that the code would assume that all the variables are dependent on one another, 
So my question is would this lead to a slow down or higher memory usage for my app? and if yes, is there a way to only read the variable value without coupling it? like passing it as a vector of variable names and use `&coupledValue()` funciton?

Thanks in advance,
Khaled

**Comment by GiudGiud** [comment, marked as answer]
Hello

If you dont couple it, then you risk the system not updating the coupled value (re-initializing it on every new quadrature point) when you try to use it. In general, if you need it you should couple it. And if you dont, you dont need to pass it in the arguments in the input file. You can always pass a constant in place of a variable if you dont plan to use that variable. 

There wont be an increased memory usage. 
There could be a slow down but it's very unlikely to be significant. As to all matters of performance, first make it work, then profile it if you need more performance. So I would not worry about this at this point. 

Guillaume

**Reply by KhaledNabilSharafeldin** [reply]
I have done some profiling, but nothing sticks out to me as out of place. Any recommendations would be really valuable. The code is both slow and consuming a lot of memory.

<details>
<summary>Memory</summary>

![pprof9079 0](https://user-images.githubusercontent.com/18222740/183485112-ed87756e-e21e-4223-9812-6348f6120512.jpg) 

</details>

<details>
<summary>CPU</summary>

![pprof16571 1](https://user-images.githubusercontent.com/18222740/183485118-0db383b0-8259-4374-a39d-4ed75491ca95.jpg)

</details>

**Reply by GiudGiud** [reply]
Seems the numerical solve is costing most of the time and the memory
Could you please paste your executioner block here?

**Reply by KhaledNabilSharafeldin** [reply]
Might also want to add that during jacobian calculation in the subapp1, memory usage doubles in size

<details>

<summary> Mainapp </summary>

```python
[Executioner]
  type = Transient

  solve_type = 'newton'
  line_search = none
     petsc_options = '-ksp_monitor'


  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  # petsc_options_iname = '-pc_type -pc_hypre_type  -pc_hypre_boomeramg_strong_threshold  -pc_hypre_boomeramg_agg_nl  -pc_hypre_boomeramg_agg_num_paths -pc_hypre_boomeramg_max_levels  -pc_hypre_boomeramg_coarsen_type  -pc_hypre_boomeramg_interp_type -pc_hypre_boomeramg_P_max -pc_hypre_boomeramg_truncfactor '
  # petsc_options_value = 'hypre    boomeramg       0.7                                   4                           5                                 25                              HIMS                              ext+i                           2                         0.3                             '
  # -pc_type hypre -pc_hypre_type boomeramg -pc_hypre_boomeramg_strong_threshold 0.7  -pc_hypre_boomeramg_agg_nl 4 -pc_hypre_boomeramg_agg_num_paths 5 -pc_hypre_boomeramg_max_levels 25 -pc_hypre_boomeramg_coarsen_type HMIS -pc_hypre_boomeramg_interp_type ext+i -pc_hypre_boomeramg_P_max 2 -pc_hypre_boomeramg_truncfactor 0.3
  l_max_its = 2
  l_tol = 1e-14
  nl_max_its = 20
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-5

  start_time = 6500.0
  # end_time = 1000000.0
  num_steps = 10
  [./TimeStepper]
    type = TransientTransport
    postprocessor = rss_max
    cfl = 0.5
    dt = 1e-6
    dt_max1 = 100
    dt_max2 = 100
    n = 12
    n_slip = 12
  [../]  
[]
```
</details>
<details>

<summary>Subapp1</summary>

```python
[Executioner]
  type = Transient
  solve_type = 'linear'                  # PJFNK JFNK NEWTON FD LINEAR
  l_tol = 1.0e-5                         # Linear Tolerance         : 1.0e-5
  l_abs_tol = 1.0e-8
  l_max_its = 500                       # Max Linear Iterations    : 10000
  # petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -pc_hypre_boomeramg_strong_threshold'
  # petsc_options_value = 'hypre    boomeramg      31                 0.7'
   petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type '
   petsc_options_value = 'asm      ilu           nonzero                  '
   
  # petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'
  # petsc_options_value = 'asm      200                ilu           NONZERO' 
  # scheme = explicit-euler
  # start_time = 55000
  # dt = 20
  [./TimeStepper]
    type = TransientTransport
    postprocessor = rss_max
    cfl = 0.45
    dt = 20
    dt_max1 = 20
    dt_max2 = 10
    n_slip = 12
    dt_time_change = 360
  [../] 
  # reset_dt = true

  # num_steps = 1000

  automatic_scaling = true
  compute_scaling_once = true 
  # end_time = 30
  # num_steps = 20
[]
```
</details>
<details>

<summary>Subapp2</summary>

```python
[Executioner]
  type = Transient
  #solve_type = 'LINEAR'                  # PJFNK JFNK NEWTON FD LINEAR
  #l_tol = 1.0e-11                         # Linear Tolerance         : 1.0e-5
  #l_max_its = 10000                       # Max Linear Iterations    : 10000
  # petsc_options_iname = '-pc_type   -pc_hypre_type  -pc_hypre_boomeramg_strong_threshold'
  # petsc_options_value = 'hypre      boomeramg       0.7'
  # start_time = 55000

  # start_time = 75

  # petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type '
  # petsc_options_value = 'asm      ilu           nonzero                  '
  # automatic_scaling = true
  # compute_scaling_once = false 
  # num_steps = 1
  # dt = 0.5
  # dt_max = 0.1
  [./TimeIntegrator]
    # type = ImplicitEuler
    # type = BDF2
    #type = CrankNicolson
    # type = ImplicitMidpoint
    # type = LStableDirk2
    # type = LStableDirk3
    # type = LStableDirk4
    # type = AStableDirk4
    #
    # Explicit methods
    type = ActuallyExplicitEuler
    # type = ExplicitEuler
    #type = ExplicitMidpoint
    # type = Heun
    # type = Ralston
  [../]
  # [./TimeStepper]
  #   type = TransientTransport
  #   postprocessor = rss_maxs
  #   cfl = 0.45
  #   dt = 1e-6
  #   dt_max = 50
  #   n_slip = 12
  # [../] 
  # scheme = implicit-euler
[]
```

</details>

**Reply by GiudGiud** [reply]
Do you get decent performance out of the subapps?

For the main app,
```
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
```
this is a direct solve and it's expensive for large systems. This is the Mat_Aij operations in the CPU profiling iirc. 
You should try a more scalable preconditioner if you want to see large gains in memory and CPU time 

For a small improvement, 
You may also try using a different solver package for LU, see if you get better performance. We have found `strumpack` to outperform the others quite often. Use the `-pc_factor_mat_solver_package` petsc option (in iname) and `strumpack` for the value.

**Reply by KhaledNabilSharafeldin** [reply]
> Do you get decent performance out of the subapps?

Yes! any suggestions for the more scalable preconditioners?

and I'll give strumpack a shot and see how that goes.
Thanks again!
Best,
Khaled

**Reply by GiudGiud** [reply]
have you tried asm with an ilu sub pc?
what about field splits? (there's a bug on schur right now, I m working on fixing it)

**Reply by KhaledNabilSharafeldin** [reply]
I think i've tried both asm and ilu, I am currently testing more with them.
However, currently I'm using a custom mesh by generating it in a main file and reading the `.e` file output. The mesh uses 2 different mesh elements PYR5 and TET. I switched this up to only PYR5 generated by moose and right away I'm getting substantially lower memory usage. I am still testing and finding the PC that works best but the mesh change made it solvable at the moment.

Do you have any reasoning why would the different meshes would make such a substantial change? mind you, they are approximately the same number of elements.

Thanks again @GiudGiud for recommendations and solutions!

Best,
Khaled

**Reply by GiudGiud** [reply]
Hard to know. Sometimes meshes are more true to the physics (think a hex mesh aligned with a fluid flow for example) and you get better solutions. I dont know about the speed of the solve though!

@roystgnr might have some insights on this

**Reply by roystgnr** [reply]
In case others come across this later, let me talk about the more common phenomena first:

Mesh changes to be more true to the physics can greatly speed up a solve on a difficult nonlinear problem by getting you to a solution in fewer quasiNewton steps - e.g. in a compressible flow, a hex mesh that's not aligned well enough with a shock can be slower to solve on or can be impossible to solve on depending on your formulation.

If you're not using either an overkill preconditioner like LU or a horribly weak preconditioner like Jacobi, I'm afraid that even seemingly trivial changes to a mesh can lead to unexpected changes in iteration counts and solve speeds.  The most runtime-efficient preconditioners take your mesh partitioning into account, with something like a Schwarz method or Block Jacobi method (for communication between processors) wrapped around a stronger preconditioner within each subpartition's diagonal block of the matrix ... and that means the strength of your preconditioner as a whole strongly depends on your partitioning.  That's a reasonable practical concern to worry about, but unfortunately it combines with a less reasonable practical concern: mesh partitioning is kind of a black art.  ParMETIS (our default for DistributedMesh) can sometimes give weird results, METIS (our default for ReplicatedMesh) can sometimes give even weirder results (despite having easier constraints on the algorithm; I don't understand why myself), and even our simple alternatives like space-filling-curves can give results very dependent on mesh geometry and processor count.  With a worse partitioning (less-"compact" partitions with more "surface" at the interfaces between them) you get a double-whammy - the reduced compactness makes your sub-preconditioners less effective so you need more iterations, and the increased surface means more communication between blocks so each iteration takes longer to boot.

But that all said ...

your problem isn't solve speed, it's solve *memory*, isn't it?  That's much more surprising to me.  There's a minor effect where going from tets to pyramids (on the same number of nodes) increases your sparsity pattern and so increases your memory requirements (both for the matrix itself and for ILU sub-preconditioners), and of course in the other direction a pyramid Mesh itself takes less memory since it's got fewer elements for the same number of nodes.  I'm not sure what else could be important.  I guess if a mesh change takes your preconditioner from "really good" to "mediocre", that can lead to an increased number of Krylov vectors being saved (by methods like GMRES, the PETSc default, which save them) and that would take more memory, but unless you're manually changing `ksp_gmres_restart` it'll throw out and reconstruct those vectors after every 30 iterations, so the memory requirements can't get out of control.

I'd love to see a before-and-after of memory usage graphs; there's got to be some factor I'm missing.

**Reply by KhaledNabilSharafeldin** [reply]
@roystgnr thank you very much for the elaborate explanation! this kind of extended replies help a lot understanding the different systems better.

The issue with my application first and foremost is memory usage, then speed. I am mostly drawing comparisons in memory usage and performance based on a similar code/model that we have written using PETSc in fortran. The ultimate goal is to get the MOOSE application up to speed and functionality of the previous one.

The mesh size I am trying to run is about 3.8-4.2 million elements, **do you have a rough estimate how much memory would be needed to run a similar mesh?** We have 3 multiapps that share the same mesh and about a 100 of Moose's "nonlinear variables", which are not necessarily nonlinear or directly coupled. Currently I've tested on a single highmemory cluster node that has 128 cores and 1TB of RAM and it still exceeded that limit.



> I think i've tried both asm and ilu, I am currently testing more with them. However, currently I'm using a custom mesh by generating it in a main file and reading the `.e` file output. The mesh uses 2 different mesh elements PYR5 and TET. I switched this up to only PYR5 generated by moose and right away I'm getting substantially lower memory usage. I am still testing and finding the PC that works best but the mesh change made it solvable at the moment.
> 
> Do you have any reasoning why would the different meshes would make such a substantial change? mind you, they are approximately the same number of elements.
> 
> Thanks again @GiudGiud for recommendations and solutions!
> 
> Best, Khaled

also regarding this, it did indeed reduce memory usage, however the increase was not as drastic as I initially seen since I got confused by `DistributedRectilinearMeshGenerator` which allows for PYRAMID type elements but instead defaults for HEX elements without any warnings or errors, so I used instead `GeneratedMeshGenerator` in distributed.

Thanks again @GiudGiud and @roystgnr!

Best,
Khaled
